### PR TITLE
[checkoutservice] add basic metric support

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

related #69.

## Changes

Try to enhance checkoutservice with metric support. 

Add basic runtime metrics for checkoutservice, using runtime metric instrumentation library.

Here is the metrics example, `process_runtime_go_goroutines`

image

***

Following paragraph is the earlier content of this pull request, but these efforts are blocked by https://github.com/open-telemetry/opentelemetry-go-contrib/pull/2700, so remove them for now.

~~According to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/rpc.md and as the first try, I first add a metrics 'rpc.server.duration' for unary service.~~

~~the metrics with four label~~
~~- rpc.system~~
~~- rpc.service~~
~~- rpc.method~~
~~- rpc.grpc.status_code~~

~~I made some experiments with new metric, I stop 'shipservice' manually to mock the fail down of shipservice and restart it after a few minutes, and draw a graph indicates the success ratio of checkout service.~~

~~the query I used is `sum(rate(rpc_server_duration_count{rpc_grpc_status_code="0"}[1m])) / sum(rate(rpc_server_duration_count[1m]))`~~

~~and the graph is like following with a drop in the middle~~

### CODE REVIEW

File: src/checkoutservice/main.go [hunk 1]
```diff
@@ -82,6 +89,43 @@ func InitTracerProvider() *sdktrace.TracerProvider {
...
+	client := otlpmetricgrpc.NewClient(
+		otlpmetricgrpc.WithInsecure())
+
+	metricExporter, err := otlpmetric.New(ctx, client)
+	if err != nil {
+		log.Errorf("failed to create the metric collector exporter, %v", err)
+		return nil, err
+	}
+
+	controller := controller.New(
```
Comment (cedricziel) [inline src/checkoutservice/main.go:104]: Is this a good point in time to introduce a shared `resource` for the controller and the tracer provider?

File: src/checkoutservice/middleware.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
+// Copyright 2018 Google LLC
```
Comment (reyang) [inline src/checkoutservice/middleware.go:1]: Do we want this or something like "Copyright The OpenTelemetry Authors"

File: src/checkoutservice/middleware.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		resp, err := handler(ctx, req)
+		if err != nil {
+			s, _ := status.FromError(err)
+			attrs = append(attrs, semconv.RPCGRPCStatusCodeKey.Int64(int64(s.Code())))
+		} else {
+			attrs = append(attrs, semconv.RPCGRPCStatusCodeOk)
+		}
+		elapsedTime := float64(time.Since(startTime)) / float64(time.Millisecond)
+		rpcServerDuration.Record(ctx, elapsedTime, attrs...)
+		fmt.Println("record meter")
```
Comment (reyang) [inline src/checkoutservice/middleware.go:75]: Remove this line?

File: src/checkoutservice/middleware.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	rpcServerDuration syncfloat64.Histogram
+)
+
+func parseFullMethod(fullMethod string) (string, []attribute.KeyValue) {
```
Comment (reyang) [inline src/checkoutservice/middleware.go:34]: Does OpenTelemetry Go provide an instrumentation library for this, or we need to manually write one? What's the best practice? (I think the goal is to show folks best practices, we don't want to show/encourage folks how to write their own HTTP protocol parser).

**Review by reyang — DISMISSED** [review]
https://github.com/open-telemetry/opentelemetry-demo/pull/339#discussion_r952814808

File: src/checkoutservice/middleware.go (hunk 4 above)
Comment (fatsheep9146) [reply src/checkoutservice/middleware.go:34]: For now, the grpc only has trace instrumentation library in go-contrib.

https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/google.golang.org/grpc/otelgrpc
image

So I tried to first add metrics support for grpc service in the demo, and in the meantime I'm trying to add metrics support for grpc in go-contrib repo, too.

What do you think is the best way? We should first add metrics support for grpc into go-contrib and then update demo to use the  instrumentation library of new version? @reyang @austinlparker

**Comment by fatsheep9146** [comment]
> [#339 (comment)](https://github.com/open-telemetry/opentelemetry-demo/pull/339#discussion_r952814808)

https://github.com/open-telemetry/opentelemetry-go-contrib/pull/2700

I also started to add metrics support for grpc instrumentation library.

**Comment by fatsheep9146** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 7 days.

Related pull requests https://github.com/open-telemetry/opentelemetry-go-contrib/pull/2700 are still being reviewed, not stale.

File: src/checkoutservice/middleware.go (hunk 2 above)
Comment (cartersocha) [reply src/checkoutservice/middleware.go:1]: Anything net new should be credited to otel!

**Comment by cartersocha** [comment]
hey @fatsheep9146 lets try to push this forward for the hero scenario asap. Our development freeze starts October 10th. 

All we need in the checkout service is a gauge on the cache size + instrumentation libraries

**Comment by cartersocha** [comment]
@austinlparker, @julianocosta89, @puckpuck fyi lets try to push this forward

File: docker-compose.yml [hunk 5]
```diff
@@ -101,6 +101,7 @@ services:
       - PRODUCT_CATALOG_SERVICE_ADDR
       - SHIPPING_SERVICE_ADDR
       - OTEL_EXPORTER_OTLP_TRACES_ENDPOINT
+      - OTEL_EXPORTER_OTLP_METRICS_ENDPOINT
```
Comment (arminru) [inline docker-compose.yml:104]:
```suggestion
      - OTEL_EXPORTER_OTLP_METRICS_ENDPOINT
      - OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE
```
see #389

**Comment by fatsheep9146** [comment]
I refract this pr to use the new version of golang metric sdk and using runtime instrumentation library. Please help review this =D 
@cartersocha @austinlparker @puckpuck @julianocosta89

File: docker-compose.yml [hunk 6]
```diff
@@ -108,6 +108,8 @@ services:
...
```
Comment (cijothomas) [inline docker-compose.yml:111]: is this related to this PR?

Comment (arminru) [reply docker-compose.yml:111]: This makes sure that the env var - if configured - is passed on to the `checkoutservice` container, where the Go SDK will automatically pick up the value to use the respective endpoint. Since this comes from the SDK OOTB, it does not involve any code changes in the implementation itself.
See https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options.

Comment (fatsheep9146) [reply docker-compose.yml:111]: Yes, this ENV is picked in the golang otel sdk

**Review by julianocosta89 — COMMENTED** [review]
I wasn't able to test this out.

Building the service is failing with the following message:
```
=> ERROR [builder 10/10] RUN go build -o /go/bin/checkoutservice/ ./                                                                    6.1s 
------                                                                                                                                        
 > [builder 10/10] RUN go build -o /go/bin/checkoutservice/ ./:                                                                               
#0 3.911 # go.opentelemetry.io/otel/sdk/metric/metricdata                                                                                     
#0 3.911 /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v0.32.1/metricdata/data.go:62:14: syntax error: unexpected int64, expecting ]        
#0 3.911 /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v0.32.1/metricdata/data.go:67:15: syntax error: unexpected ), expecting type         
#0 3.911 /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v0.32.1/metricdata/data.go:70:12: syntax error: unexpected int64, expecting ]        
#0 3.911 /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v0.32.1/metricdata/data.go:80:13: syntax error: unexpected ), expecting type
#0 3.911 /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v0.32.1/metricdata/data.go:83:18: syntax error: unexpected int64, expecting ]
#0 3.911 note: module requires Go 1.18
------
failed to solve: executor failed running [/bin/sh -c go build -o /go/bin/checkoutservice/ ./]: exit code: 2
```

I've tried updating the go version within the `Dockerfile`and `go.mod`, but then the build failed with the following:
```
=> ERROR [builder  9/10] RUN protoc -I ./proto/ ./proto/demo.proto --go_out=./ --go-grpc_out=./                                         0.4s 
------                                                                                                                                        
 > [builder  9/10] RUN protoc -I ./proto/ ./proto/demo.proto --go_out=./ --go-grpc_out=./:
#0 0.412 protoc-gen-go: program not found or is not executable
#0 0.412 Please specify a program using absolute path or make sure the program is available in your PATH system variable
#0 0.412 --go_out: protoc-gen-go: Plugin failed with status code 1.
------
failed to solve: executor failed running [/bin/sh -c protoc -I ./proto/ ./proto/demo.proto --go_out=./ --go-grpc_out=./]: exit code: 1
```

**Comment by austinlparker** [comment]
I'm investigating this - looks like part of it is that the go version in the container isn't 1.18+

**Comment by julianocosta89** [comment]
@austinlparker yes, but when I tried updating the go version within the `Dockerfile` and `go.mod`, the build failed with the following:
```
=> ERROR [builder  9/10] RUN protoc -I ./proto/ ./proto/demo.proto --go_out=./ --go-grpc_out=./                                                       0.6s
------
 > [builder  9/10] RUN protoc -I ./proto/ ./proto/demo.proto --go_out=./ --go-grpc_out=./:
#0 0.552 protoc-gen-go: program not found or is not executable
#0 0.552 Please specify a program using absolute path or make sure the program is available in your PATH system variable
#0 0.553 --go_out: protoc-gen-go: Plugin failed with status code 1.
------
failed to solve: executor failed running [/bin/sh -c protoc -I ./proto/ ./proto/demo.proto --go_out=./ --go-grpc_out=./]: exit code: 1
```

**Comment by austinlparker** [comment]
@fatsheep9146 it looks like in order to take newer versions of otel-go we will need to upgrade all go services to support 1.18+. I believe we will also need to update to the newer protobuf package. Can you take care of that for this PR as well?

**Comment by fatsheep9146** [comment]
> @fatsheep9146 it looks like in order to take newer versions of otel-go we will need to upgrade all go services to support 1.18+. I believe we will also need to update to the newer protobuf package. Can you take care of that for this PR as well?

Yes, I already found this problem and started doing it. I also found that in 1.18, `go get` can would not install the package, so the  dockerfile should also be updated. @austinlparker

**Comment by austinlparker** [comment]
> > @fatsheep9146 it looks like in order to take newer versions of otel-go we will need to upgrade all go services to support 1.18+. I believe we will also need to update to the newer protobuf package. Can you take care of that for this PR as well?
> 
> 
> 
> Yes, I already found this problem and started doing it. I also found that in 1.18, `go get` can would not install the package, so the  dockerfile should also be updated. @austinlparker 

Yeah - I would just upgrade to 1.19 across the board, as well as switch to the newer Google protobuf package. Maybe best to do that in a new PR just for cleanliness sake.

**Comment by fatsheep9146** [comment]
> > > @fatsheep9146 it looks like in order to take newer versions of otel-go we will need to upgrade all go services to support 1.18+. I believe we will also need to update to the newer protobuf package. Can you take care of that for this PR as well?
> > 
> > 
> > Yes, I already found this problem and started doing it. I also found that in 1.18, `go get` can would not install the package, so the  dockerfile should also be updated. @austinlparker
> 
> Yeah - I would just upgrade to 1.19 across the board, as well as switch to the newer Google protobuf package. Maybe best to do that in a new PR just for cleanliness sake.

Ok, I will first submit a new pull request to update the dockerfile and protobuf version. After that pr is merged I will rebase this pr then.

**Comment by julianocosta89** [comment]
Hello all,

I've just sent a PR with the required update: #406 

Once, this PR is merged we can simply rebase and merge this one.
Already tested locally and it is working 🥳

**Review by mic-max — APPROVED** [review]

**Comment by cartersocha** [comment]
@fatsheep9146 the go version is now updated!

**Comment by fatsheep9146** [comment]
> @fatsheep9146 the go version is now updated! thanks @julianocosta89

I also tested it locally, it worked, thanks! @cartersocha @julianocosta89

File: src/checkoutservice/main.go (hunk 1 above)
Comment (fatsheep9146) [reply src/checkoutservice/main.go]: I think this may be better to be solved in another pr, since this also changed the behavior of trace?
@cedricziel @cartersocha

**Review by julianocosta89 — APPROVED** [review]
LGTM

**Comment by julianocosta89** [comment]
@reyang anything missing in here?

**Comment by reyang** [comment]
> @reyang anything missing in here?

Let me take a quick look.

**Review by reyang — APPROVED** [review]
LGTM.

**Comment by reyang** [comment]
I don't know if this needs a changelog entry or not. Not a blocker though.

**Comment by cartersocha** [comment]
I'll cover the changelog update in a separate PR to move this forward

File: src/checkoutservice/main.go (hunk 1 above)
Comment (cartersocha) [reply src/checkoutservice/main.go]: resolving this for now. We can discuss more over slack
